rlinf/hybrid_engines/megatron/megatron_model_manager.py

```diff
@@ -447,17 +480,44 @@ def offload_model_weights_and_grad(self, offload_grad=True, offload_weight=True)
             else:
                 for param_name, param in model_chunk.named_parameters():
                     if offload_weight and param.data is not None:
-                        param.data = param.data.to("cpu", non_blocking=True)
+                        cpu_data = self._get_pinned_buffer(param)
+                        cpu_data.copy_(param.data, non_blocking=True)
+                        param.data = cpu_data
 
                     if offload_grad and param.grad is not None:
-                        param.grad = param.grad.to("cpu", non_blocking=True)
+                        # For gradients, we attach the pinned buffer to the param object
+                        # because param.grad tensor might be recreated by autograd
+                        if (
+                            not hasattr(param, "cpu_grad_data")
+                            or param.cpu_grad_data is None
+                        ):
+                            param.cpu_grad_data = torch.empty(
+                                param.grad.size(),
+                                dtype=param.grad.dtype,
+                                layout=param.grad.layout,
+                                pin_memory=True,
+                                device="cpu",
+                            )
+                        param.cpu_grad_data.copy_(param.grad, non_blocking=True)
+                        param.grad = param.cpu_grad_data
+
+        if buffers_to_resize:
```

Comment:
When the model is not DDP (although I'm not sure when will this happen in Megatron), non of the tensors will be added to `buffers_to_resize`, leading to missing `cuda.synchronize`. I think this is will cause data corruption when using pin memory. See https://docs.pytorch.org/tutorials/intermediate/pinmem_nonblock.html#other-copy-directions-gpu-cpu-cpu-mps. Please call synchronize in any case.